File: python/tests/test_LocalBrowserLaunchExtensionInjector.py
from __future__ import annotations

import unittest
import time
import tempfile
import zipfile
from pathlib import Path
from typing import Any, cast

from modcdp.injector.ExtensionInjector import DEFAULT_MODCDP_EXTENSION_ID
from modcdp.injector.LocalBrowserLaunchExtensionInjector import LocalBrowserLaunchExtensionInjector


ROOT = Path(__file__).resolve().parents[2]
EXTENSION_PATH = ROOT / "dist" / "extension"


class LocalBrowserLaunchExtensionInjectorTests(unittest.TestCase):
    def test_rejects_zip_entries_outside_extraction_directory(self) -> None:
        with tempfile.TemporaryDirectory(prefix="modcdp-bad-zip-") as temp_dir:
            zip_path = Path(temp_dir) / "extension.zip"
            with zipfile.ZipFile(zip_path, "w") as archive:
                archive.writestr("../evil.txt", "evil")

            injector = LocalBrowserLaunchExtensionInjector({"injector_extension_path": str(zip_path)})
            try:
                with self.assertRaisesRegex(RuntimeError, "escapes extension extraction directory"):
                    injector.prepare()
                self.assertFalse((Path(temp_dir) / "evil.txt").exists())
            finally:
                injector.close()

    def test_prepares_unpacked_extension_directory_for_load_extension(self) -> None:
        injector = LocalBrowserLaunchExtensionInjector({"injector_extension_path": str(EXTENSION_PATH)})
        try:
            injector.prepare()
            unpacked_extension_path = injector.unpacked_extension_path
            self.assertIsInstance(unpacked_extension_path, str)
            unpacked_extension_path = cast(str, unpacked_extension_path)
            self.assertNotEqual(unpacked_extension_path, str(EXTENSION_PATH))
            self.assertTrue((Path(unpacked_extension_path) / "manifest.json").exists())
            self.assertEqual(injector.getLauncherConfig(), {"extra_args": [f"--load-extension={unpacked_extension_path}"]})
            self.assertEqual(injector.options.get("injector_extension_id"), DEFAULT_MODCDP_EXTENSION_ID)
        finally:
            injector.close()

    def test_prepares_default_extension_zip_for_load_extension(self) -> None:
        injector = LocalBrowserLaunchExtensionInjector()
        try:
            injector.prepare()
            unpacked_extension_path = injector.unpacked_extension_path
            self.assertIsInstance(unpacked_extension_path, str)
            unpacked_extension_path = cast(str, unpacked_extension_path)
            self.assertTrue((Path(unpacked_extension_path) / "manifest.json").exists())
            self.assertIn("modcdp-extension-", unpacked_extension_path)
            self.assertEqual(injector.getLauncherConfig(), {"extra_args": [f"--load-extension={unpacked_extension_path}"]})
            self.assertEqual(injector.options.get("injector_extension_id"), DEFAULT_MODCDP_EXTENSION_ID)
        finally:
            injector.close()

    def test_returns_immediately_when_launched_extension_target_is_absent(self) -> None:
        methods: list[str] = []

        def send(method: str, params: dict[str, Any] | None = None, session_id: str | None = None) -> dict[str, Any]:
            methods.append(method)
            if method == "Target.getTargets":
                return {"targetInfos": []}
            raise RuntimeError(f"unexpected {method}")

        injector = LocalBrowserLaunchExtensionInjector(
            cast(Any, {
                "injector_extension_path": str(EXTENSION_PATH),
                "injector_trust_service_worker_target": True,
                "send": send,
            })
        )
        try:
            injector.prepare()
            started_at = time.perf_counter()
            result = injector.inject()
            elapsed_ms = (time.perf_counter() - started_at) * 1000
            self.assertIsNone(result)
            self.assertEqual(methods, ["Target.getTargets"])
            self.assertLess(elapsed_ms, 200)
        finally:
            injector.close()


if __name__ == "__main__":
    unittest.main()

File: js/test/test.LocalBrowserLaunchExtensionInjector.ts
import assert from "node:assert/strict";
import { existsSync, mkdtempSync, rmSync, writeFileSync } from "node:fs";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { test } from "vitest";

import { DEFAULT_MODCDP_EXTENSION_ID } from "../src/injector/ExtensionInjector.js";
import { LocalBrowserLaunchExtensionInjector } from "../src/injector/LocalBrowserLaunchExtensionInjector.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const EXTENSION_PATH = path.resolve(HERE, "..", "..", "dist", "extension");

function crc32(data: Buffer) {
  let crc = 0xffffffff;
  for (const byte of data) {
    crc ^= byte;
    for (let bit = 0; bit < 8; bit++) crc = (crc >>> 1) ^ (0xedb88320 & -(crc & 1));
  }
  return (crc ^ 0xffffffff) >>> 0;
}

function storedZipEntry(name: string, data: Buffer) {
  const name_buffer = Buffer.from(name);
  const checksum = crc32(data);
  const local = Buffer.alloc(30);
  local.writeUInt32LE(0x04034b50, 0);
  local.writeUInt16LE(20, 4);
  local.writeUInt32LE(checksum, 14);
  local.writeUInt32LE(data.length, 18);
  local.writeUInt32LE(data.length, 22);
  local.writeUInt16LE(name_buffer.length, 26);
  const central = Buffer.alloc(46);
  central.writeUInt32LE(0x02014b50, 0);
  central.writeUInt16LE(20, 4);
  central.writeUInt16LE(20, 6);
  central.writeUInt32LE(checksum, 16);
  central.writeUInt32LE(data.length, 20);
  central.writeUInt32LE(data.length, 24);
  central.writeUInt16LE(name_buffer.length, 28);
  const central_offset = local.length + name_buffer.length + data.length;
  const end = Buffer.alloc(22);
  end.writeUInt32LE(0x06054b50, 0);
  end.writeUInt16LE(1, 8);
  end.writeUInt16LE(1, 10);
  end.writeUInt32LE(central.length + name_buffer.length, 12);
  end.writeUInt32LE(central_offset, 16);
  return Buffer.concat([local, name_buffer, data, central, name_buffer, end]);
}

test("LocalBrowserLaunchExtensionInjector rejects zip entries outside extraction directory", async () => {
  const temp_dir = mkdtempSync(path.join(os.tmpdir(), "modcdp-bad-zip-"));
  const zip_path = path.join(temp_dir, "extension.zip");
  writeFileSync(zip_path, storedZipEntry("../evil.txt", Buffer.from("evil")));
  const injector = new LocalBrowserLaunchExtensionInjector({
    injector_extension_path: zip_path,
  });

  try {
    await assert.rejects(() => injector.prepare(), /escapes extension extraction directory/);
    assert.equal(existsSync(path.join(temp_dir, "evil.txt")), false);
  } finally {
    await injector.close();
    rmSync(temp_dir, { recursive: true, force: true });
  }
});

test("LocalBrowserLaunchExtensionInjector prepares an unpacked extension directory for --load-extension", async () => {
  const injector = new LocalBrowserLaunchExtensionInjector({
    injector_extension_path: EXTENSION_PATH,
  });

  try {
    await injector.prepare();
    const unpacked_extension_path = (injector as unknown as { unpacked_extension_path?: string | null })
      .unpacked_extension_path;
    assert.equal(typeof unpacked_extension_path, "string");
    assert.notEqual(unpacked_extension_path, EXTENSION_PATH);
    assert.equal(existsSync(path.join(unpacked_extension_path!, "manifest.json")), true);
    assert.deepEqual(injector.getLauncherConfig(), {
      extra_args: [`--load-extension=${unpacked_extension_path}`],
    });
    assert.equal(injector.options.injector_extension_id, DEFAULT_MODCDP_EXTENSION_ID);
  } finally {
    await injector.close();
  }
});

test("LocalBrowserLaunchExtensionInjector prepares the default extension zip for --load-extension", async () => {
  const injector = new LocalBrowserLaunchExtensionInjector();

  try {
    await injector.prepare();
    const unpacked_extension_path = (injector as unknown as { unpacked_extension_path?: string | null })
      .unpacked_extension_path;
    assert.equal(typeof unpacked_extension_path, "string");
    assert.match(unpacked_extension_path!, /modcdp-extension-/);
    assert.equal(existsSync(path.join(unpacked_extension_path!, "manifest.json")), true);
    assert.deepEqual(injector.getLauncherConfig(), {
      extra_args: [`--load-extension=${unpacked_extension_path}`],
    });
    assert.equal(injector.options.injector_extension_id, DEFAULT_MODCDP_EXTENSION_ID);
  } finally {
    await injector.close();
  }
});

test("LocalBrowserLaunchExtensionInjector returns immediately when the launched extension target is absent", async () => {
  const methods: string[] = [];
  const injector = new LocalBrowserLaunchExtensionInjector({
    injector_extension_path: EXTENSION_PATH,
    injector_trust_service_worker_target: true,
    send: async (method) => {
      methods.push(method);
      if (method === "Target.getTargets") return { targetInfos: [] };
      throw new Error(`unexpected ${method}`);
    },
  });

  try {
    await injector.prepare();
    const started_at = performance.now();
    const result = await injector.inject();
    const elapsed_ms = performance.now() - started_at;
    assert.equal(result, null);
    assert.deepEqual(methods, ["Target.getTargets"]);
    assert.equal(elapsed_ms < 200, true, `inject() took ${elapsed_ms}ms`);
  } finally {
    await injector.close();
  }
});
